datapath: Use bpftool for generating BPF feature macros

vadorovsky opened the pull request:
To adjust BPF programs to different kernels containing different BPF features, bpftool provides macros like i.e.:

```
# bpftool feature probe macros
/*** System call availability ***/
#define HAVE_BPF_SYSCALL
[...]
/*** eBPF map types ***/
#define HAVE_HASH_MAP_TYPE
[...]
```

Cilium implemented similar macros on its own, but this pull request removes them in favor of macros generated by bpftool.

```release-note
Use bpftool for generating BPF feature macros
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10019)
<!-- Reviewable:end -->

vadorovsky committed:
datapath: Use bpftool for generating BPF feature macros

To adjust BPF programs to different kernels containing different BPF
features, bpftool provides macros like i.e.:

  # bpftool feature probe macros
  /*** System call availability ***/
  #define HAVE_BPF_SYSCALL
  [...]
  /*** eBPF map types ***/
  #define HAVE_HASH_MAP_TYPE
  [...]

Cilium implemented similar macros on its own, but from this change it's
going to use the bpftool implementation. After this change, the output
of `bpftool feature probe macros` is written to the `bpf_features.h`.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
bpf: Remove old style probes

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
bpf: Use bpftool feature macros

Previous changed dropped our custom BPF feature probes in favor of
bpftool which also generates feature macros for BPF programs. This
change switches all BPF programs to them.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

pchaigno commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -217,3 +223,60 @@ func (p *ProbeManager) GetHelpers(prog string) map[string]struct{} {
+		stderr, err := ioutil.ReadAll(stderrPipe)
+		if err != nil {
+			return fmt.Errorf(
+				"bpftool did not run succesfully and stdout could not be read: %s",
```
Typo on `successfully` here. Maybe this could be shorten to `bpftool failed: %s` though.

pchaigno commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -217,3 +223,60 @@ func (p *ProbeManager) GetHelpers(prog string) map[string]struct{} {
+	}
+	defer featuresFile.Close()
+
+	if err := p.writeHeaders(featuresFile); err != nil {
```
I might have missed it, but it doesn't look like you preserve the file's header and footer, i.e., `#ifndef BPF_FEATURES_H_` & co.

vadorovsky replied:
Thanks for pointing that out! Fixed, I added header and footer in `writeHeaders`.

pchaigno commented on pkg/datapath/linux/probes/probes_privileged_test.go:
```diff
@@ -48,3 +50,26 @@ func (s *ProbesPrivTestSuite) TestHelpers(c *C) {
+	content := buf.String()
+
+	for _, pattern := range expectedPatterns {
+		c.Assert(content, Matches, pattern)
```
Unfortunately, gocheck's `Matches` [requires regular expressions that match the full string](https://github.com/adjust/gocheck/blob/fbc315b36e0e7dc95023e96b43af5d8de58bc6fb/checkers.go#L307). A first solution would be to frame `pattern` with match-all statements:
```go
c.Assert(content, Matches, "(.|\\n)*"+pattern+"(.|\\n)*")
```
But that's very ugly. A better solution is probably to define our own `PartialMatches` in `pkg/checker/checker.go` to allow matching against a subset of the given string. That would look something like that:
```go
type matchesChecker struct {
	*check.CheckerInfo
}

// PartialMatches is a GoCheck checker that the provided regex matches at least
// part of the provided string. It can act as a substitute for Matches.
var (
	matchesParams                = []string{"value", "regex"}
	PartialMatches check.Checker = &matchesChecker{
		&check.CheckerInfo{Name: "PartialMatches", Params: matchesParams},
	}
)

func (checker *matchesChecker) Check(params []interface{}, _ []string) (result bool, error string) {
	return matches(params[0], params[1])
}

func matches(value, regex interface{}) (result bool, error string) {
	regexStr, ok := regex.(string)
	if !ok {
		return false, "Regex must be a string"
	}
	valueStr, ok := value.(string)
	if !ok {
		return false, "Value must be a string"
	}
	matches, err := regexp.MatchString(regexStr, valueStr)
	if err != nil {
		return false, "Failed to compile regex: " + err.Error()
	}
	return matches, ""
}
```

pchaigno reviewed (commented):
Hey @mrostecki :wave:  This is great!

I have a comment below that might address (one of) the test failures. I also found a couple comments in the Go code that would need to be updated:
```
$ git grep -P "(HAVE_LPM_MAP_TYPE|HAVE_SKB_CHANGE_TAIL|HAVE_GET_SOCK_COOKIE|HAVE_LRU_MAP_TYPE|HAVE_FIB_LOOKUP)"
Documentation/contributing/development/dev_setup.rst:    HAVE_LPM_MAP_TYPE: Your kernel doesn't support LPM trie maps for BPF, thus disabling CIDR policies. Recommendation is to run 4.11+ kernels.
Documentation/contributing/development/dev_setup.rst:    HAVE_LRU_MAP_TYPE: Your kernel doesn't support LRU maps for BPF, thus switching back to using hash table for the cilium connection tracker. Recommendation is to run 4.10+ kernels.
pkg/bpf/map_linux.go:// the same probe logic (eg HAVE_LRU_MAP_TYPE, HAVE_LPM_MAP_TYPE).
```

vadorovsky committed:
doc: Remove information about probes logs in /var/run/cilium/state

Previous changes remove Cilium BPF feature probes in favor of bpftool
probes, which are executed by daemon in Go code and errors are included
in the cilium-agent log. Additional .log file is not created anymore.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
checker: Support partial regexp matches

Matches from gocheck doesn't support partial regexp matches and requires
expressions that match the full string. This change introduces the new
PartialMatches checker which solves that problem.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

pchaigno commented on bpf/bpf_features.h:
```diff
@@ -1,6 +1,6 @@
 #define BPF_FEATURES_H_
 
-#define HAVE_SKB_CHANGE_TAIL
+#define BPF__PROG_TYPE_sched_cls__HELPER_bpf_skb_change_tail
```
This needs to be `#define BPF__PROG_TYPE_sched_cls__HELPER_bpf_skb_change_tail 1` to enable tail calls by default when building BPF programs.

vadorovsky replied:
I also added the following macro from bpftool there:

```
#define HAVE_PROG_TYPE_HELPER(prog_type, helper)        \
        BPF__PROG_TYPE_ ## prog_type ## __HELPER_ ## helper
```

pchaigno commented on bpf/bpf_netdev.c:
```diff
@@ -507,7 +507,7 @@ static __always_inline int do_netdev_encrypt_fib(struct __ctx_buff *ctx,
 	int ret = 0;
 
-#ifdef HAVE_FIB_LOOKUP
+#ifdef BPF__PROG_TYPE_sched_cls__HELPER_bpf_fib_lookup
```
Ok, so the tests are currently failing because bpftool always defines these helper macros. They have value 1 or 0 depending on whether the helper is available or not. So this line and other lines checking for helper availability need to be changed to use `#if` instead of `#ifdef`. The same is not true for the `HAVE_` macros which bpftool does not define if the feature isn't available.

vadorovsky replied:
I also discovered that bpftool provides the following macro:

```
/*** eBPF helper functions ***/
/*
 * Use HAVE_PROG_TYPE_HELPER(prog_type_name, helper_name)
 * to determine if <helper_name> is available for <prog_type_name>,
 * e.g.
 *      #if HAVE_PROG_TYPE_HELPER(xdp, bpf_redirect)
 *              // do stuff with this helper
 *      #elif
 *              // use a workaround
 *      #endif
 */
#define HAVE_PROG_TYPE_HELPER(prog_type, helper)        \
        BPF__PROG_TYPE_ ## prog_type ## __HELPER_ ## helper
```

I used it now.

pchaigno commented on pkg/datapath/linux/probes/probes_privileged_test.go:
```diff
@@ -48,3 +52,26 @@ func (s *ProbesPrivTestSuite) TestHelpers(c *C) {
+		"#define HAVE.*PROG_TYPE",
+		"#define HAVE.*MAP_TYPE",
+		"#define HAVE.*HELPER",
+		"#endif /\\* BPF_FEATURES_H_ \\*/",
```
Maybe add another regular expression here to match the `#define BPF__PROG_TYPE_.*__HELPER_.* [01]` cases?

pchaigno reviewed (approved):
The Jenkins build has a single failure and it's known flake #10565. The Travis CI build failed for aarch64, I restarted it.

EDIT: The Travis CI build for aarch64 passed on the second run.

LGTM otherwise!

joestringer commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -217,3 +223,66 @@ func (p *ProbeManager) GetHelpers(prog string) map[string]struct{} {
+// features are available in the kernel.
+func (p *ProbeManager) CreateHeadersFile() error {
+	globalsDir := option.Config.GetGlobalsDir()
+	if err := os.MkdirAll(globalsDir, defaults.StateDirRights); err != nil && os.IsExist(err) {
```
This error check doesn't seem right.

From the `os.MkdirAll()` documentation:

```
If path is already a directory, MkdirAll does nothing and returns nil.
```

Here we seem to be only checking for one error type and failing only on that error type.

I think we can just do this?
```suggestion
	if err := os.MkdirAll(globalsDir, defaults.StateDirRights); err != nil {
```

joestringer commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -217,3 +223,66 @@ func (p *ProbeManager) GetHelpers(prog string) map[string]struct{} {
+	defer featuresFile.Close()
+
+	if err := p.writeHeaders(featuresFile); err != nil {
+		return err
```
nit: I wonder if it's worth wrapping this error here to make it clear that the `bpftool` failure was during features header write (rather than eg. sockmap load). All of the errors created in the `writeHeaders()` call seem to just talk about failures running bpftool so it might be useful to have this small extra piece of info.

You might consider using `github.com/pkg/errors` for this if you think it's useful, I think we pulled that dependency in elsewhere recently.

vadorovsky replied:
I like your idea, I will do that.

vadorovsky replied:
I didn't wrap that particular err you commented here, but I wrapped all the others returned in `writeHeaders` and `CreateHeadersFile`.

joestringer commented on pkg/datapath/linux/requirements.go:
```diff
@@ -203,23 +203,15 @@ func CheckMinRequirements() {
-	}
-
-	checkBPFLogs("bpf_requirements", true)
-	checkBPFLogs("bpf_features", false)
```
I think this means we can also remove the `checkBPFLogs()` function.

joestringer replied:
(Are all of the existing checks from that function already handled in `SystemConfigProbes()` or elsewhere?)

vadorovsky replied:
> (Are all of the existing checks from that function already handled in `SystemConfigProbes()` or elsewhere?)

`bpf_requirements` checks are handled in `SystemConfigProbes()`, `bpf_features` checks in `CreateHeadersFile`.

joestringer reviewed (changes requested):
Anticipating this landing soon :D :rocket: 

A few minor bits of feedback below, in particular there's one error check that I couldn't quite understand.
